### PR TITLE
fix: TokenUsage stats is not calculated properly for apps when Core have multiple instances #393

### PR DESCRIPTION
**Pull request by astsiapanay**

### CODE REVIEW

File: src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        deploymentStats.deploymentCostStats = deploymentCostStats;
+        String parenSpanId = deploymentStats.parentSpanId;
+        while (parenSpanId != null) {
+            deploymentStats = traceContext.spans.get(parenSpanId);
+            deploymentStats.deploymentCostStats.increase(deploymentCostStats);
+            parenSpanId = deploymentStats.parentSpanId;
+        }
+    }
+
+    private boolean isStreamingResponseCompleted(Buffer chunk, int index) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java]: consider move this method to the ProxyUtil next to isStreamingResponse method

File: src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            TraceContext traceContext = ProxyUtil.convertToObject(json, TraceContext.class);
+            if (traceContext == null) {
+                traceContext = new TraceContext();
+            }
+            traceContext.addSpan(context);
+            return ProxyUtil.convertToString(traceContext);
+        });
+    }
+
+    public Future<DeploymentCostStats> getDeploymentStats(ProxyContext context) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java:58]: can we refactor this method to
```java
    public Future<DeploymentCostStats> getDeploymentStats(ProxyContext context) {
        return vertx.executeBlocking(() -> {
            ResourceDescription resource = toResource(context.getTraceId());
            String json = resourceService.getResource(resource);
            TraceContext traceContext = ProxyUtil.convertToObject(json, TraceContext.class);
            if (traceContext == null) {
                return null;
            }
            return traceContext.getStats(context);
        }, false);
    }
```

File: src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+                    return null;
+                }
+                return traceContext.getStats(context);
+            }, false);
+        } catch (Throwable e) {
+            return Future.failedFuture(e);
+        }
+    }
+
+    public Future<Void> endSpan(ProxyContext context) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java:73]: can we refactor this method to
```java
public Future<Void> endSpan(ProxyContext context) {
            ApiKeyData apiKeyData = context.getApiKeyData();
            if (apiKeyData.getPerRequestKey() != null) {
                return Future.succeededFuture();
            }
               
            return vertx.executeBlocking(() -> {
                ResourceDescription resource = toResource(context.getTraceId());
                resourceService.deleteResource(resource);
                return null;
            }, false);
    }
```

File: src/main/java/com/epam/aidial/core/function/CollectModelStatsFn.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import com.epam.aidial.core.ProxyContext;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+public class CollectModelStatsFn extends BaseFunction<ObjectNode> {
+    public CollectModelStatsFn(Proxy proxy, ProxyContext context) {
+        super(proxy, context);
+    }
+
+    @Override
+    public Throwable apply(ObjectNode requestBody) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/function/CollectModelStatsFn.java:19]: please add a javadoc or comment with description why we do not catch error from this code and return null in any case

File: src/main/java/com/epam/aidial/core/log/GfLogStore.java [hunk 5]
```diff
@@ -77,7 +78,8 @@ private void append(ProxyContext context, LogEntry entry) throws JsonProcessingE
         append(entry, request.getHeader(Proxy.HEADER_JOB_TITLE), true);
         append(entry, "\"}", false);
 
-        TokenUsage tokenUsage = context.getTokenUsage();
+        DeploymentCostStats deploymentCostStats = context.getDeploymentCostStats();
+        TokenUsage tokenUsage = deploymentCostStats.getTokenUsage();
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/log/GfLogStore.java]: this may lead to NPE in DeploymentFeatureController

Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/log/GfLogStore.java]: fixed

File: src/main/java/com/epam/aidial/core/function/CollectModelStatsFn.java (hunk 4 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/function/CollectModelStatsFn.java:19]: fixed

File: src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java (hunk 3 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java:73]: The method should always return an instance of future even in the case of error.

File: src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java (hunk 2 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java:58]: The method should always return an instance of future even in the case of error.

File: src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java (hunk 1 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java]: fixed

File: src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java (hunk 2 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java:58]: fixed

File: src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java (hunk 3 above)
Comment (astsiapanay) [reply src/main/java/com/epam/aidial/core/token/DeploymentCostStatsTracker.java:73]: fixed

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-398.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-398.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-398.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-398.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by astsiapanay** [comment]
The main issue is that we have to assemble streaming response using individual chunks. That adds a complexity to the solution.
The decision was made in favor of https://github.com/epam/ai-dial-core/pull/402

**Pull request closed without merging**